lib/pbench/server/auth/__init__.py

```diff
@@ -62,10 +61,9 @@ def __init__(
         self.headers = CaseInsensitiveDict([] if headers is None else headers)
         self.verify = verify
         self.connection = requests.session()
-        # Note: At this point the OpenIDClient object variables initialization
-        # is done, and we start initializing the class variables.
+        self.public_key = self._get_oidc_public_key()
+        # TODO: Think about a better way to set well known endpoints
```

Comment:
We still have the issue that we're calling an instance method from inside the instance's constructor.  We should have that big "caution" comment here warning those who come after that before line 64 is the object initialization and what comes after it is something else, so that they don't add something in the wrong place.

Perhaps the solution would be for this class to provide a factory method which could be called to instantiate and initialize the class and instance; then, instead of calling the constructor, anybody who wants an `OpenIDClient` object could call the factory function, and that function could call the constructor and then use the resulting object to set the cached key value and the well-known endpoints.  It could encapsulate the settings so that they only get done once, and it could even reuse the instance if we want it to be a singleton.

Or, maybe we just need to get rid of this class too...do we really need to instantiate this object?  Do we need to do so more than once?  If not, then we probably don't need a class here.

Reply:
I created an issue https://github.com/distributed-system-analysis/pbench/issues/3102, I will address it as a separate PR.

Reply:
The change proposed in this PR is to add aan invocation of an instance method inside the constructor for that object.  This is a highly questionable maneuver, and so, at minimum, this code change should include a warning comment in the code.

I'll grant that you've inserted this new call right before an existing call to another instance method, so, yes, the problem already existed, but that's not a justification for not adding the appropiate code comment.  

The particular problem with the new code line is that it appears to be simply assigning a value to an instance attribute; someone refactoring or "cleaning up" the `__init__()` function might think that they can re-order the attribute initializations, but the call to `_get_oidc_public_key()` hides the fact that it depends on the previous initializations...hence the need for a big, scary code comment.

We can use #3102 to track implementing the _proper_ fix here, but, in the meantime, we should be carrying a suitable warning in the code.

Reply:
A *big* warning. Note that the method we're calling depends both on `self.realm_name` being set *plus* any attribute dependencies of `self._get()`, which are another layer deeper and correspondingly more awkward to discover. This is fragile and arcane, adding a lot of risk to the code.

Reply:
[See below](https://github.com/distributed-system-analysis/pbench/pull/3095#discussion_r1039987739) for a superb, immediate solution.